src/beignet/func/_partition.py

```diff
@@ -0,0 +1,1727 @@
+        An object representing partition errors.
+    reference_positions : Tensor
+        A tensor containing the reference positions of particles.
+    units_buffer_size : int or None
+        The buffer size in units.
+    update_fn : Callable[[Tensor, "_NeighborList", Any], "_NeighborList"]
+        A function to update the neighbor list.
+
+    Methods
+    -------
+    update(positions: Tensor, **kwargs) -> "_NeighborList"
+        Updates the neighbor list with new positions.
+    did_buffer_overflow() -> bool
+        Checks if the buffer overflowed.
+    cell_size_too_small() -> bool
+        Checks if the cell size is too small.
+    malformed_box() -> bool
+        Checks if the box is malformed.
+    """
+
+    buffer_fn: Callable[[Tensor, _CellList], _CellList] = static_field()
+    indexes: Tensor
+    item_size: float | None = static_field()
+    maximum_size: int = static_field()
+    format: _NeighborListFormat = static_field()
+    partition_error: _PartitionError
+    reference_positions: Tensor
+    units_buffer_size: int | None = static_field()
+    update_fn: Callable[[Tensor, "_NeighborList", Any], "_NeighborList"] = (
+        static_field()
+    )
+
+    def update(self, positions: Tensor, **kwargs) -> "_NeighborList":
+        return self.update_fn(positions, self, **kwargs)
+
+    @property
+    def did_buffer_overflow(self) -> bool:
+        return (
+            self.partition_error.code
+            & (PEC.NEIGHBOR_LIST_OVERFLOW | PEC.CELL_LIST_OVERFLOW)
+        ).item() != 0
+
+    @property
+    def cell_size_too_small(self) -> bool:
+        return (self.error.code & PEC.CELL_SIZE_TOO_SMALL).item() != 0
+
+    @property
+    def malformed_box(self) -> bool:
+        return (self.error.code & PEC.MALFORMED_BOX).item() != 0
+
+
+@_dataclass
+class _NeighborListFunctionList:
+    r"""
+    A dataclass that encapsulates functions for setting up and updating a neighbor list.
+
+    Attributes
+    ----------
+    setup_fn : Callable[..., _NeighborList]
+        A function that sets up and returns a `_NeighborList` object.
+    update_fn : Callable[[Tensor, _NeighborList], _NeighborList]
+        A function that updates a `_NeighborList` object given a tensor and an existing `_NeighborList`.
+
+    Methods
+    -------
+    __iter__()
+        Returns an iterator over the setup and update functions.
+    """
+
+    setup_fn: Callable[..., _NeighborList] = static_field()
+
+    update_fn: Callable[[Tensor, _NeighborList], _NeighborList] = static_field()
+
+    def __iter__(self):
+        return iter((self.setup_fn, self.update_fn))
+
+
+def distance(dR: Tensor) -> Tensor:
+    """Computes distances.
+
+    Args:
+      dR: Matrix of displacements; `Tensor(shape=[..., spatial_dim])`.
+    Returns:
+      Matrix of distances; `Tensor(shape=[...])`.
+    """
+    dr = square_distance(dR)
+
+    return safe_mask(dr > 0, torch.sqrt, dr)
+
+
+def _iota(shape: tuple[int, ...], dim: int = 0, **kwargs) -> Tensor:
```

Comment:
Very nice. Should be public and moved to root since it's so useful and PyTorch does not provide an iota.

Reply:
Should also be tested against numpy.iota.